spicy/toolchain/src/compiler/codegen/parsers/types.cc

```diff
@@ -342,24 +448,15 @@ Expression* TypeParser::buildParser(UnqualifiedType* t) {
 
 } // namespace
 
-Expression* ParserBuilder::_parseType(UnqualifiedType* t, const production::Meta& meta, Expression* dst, bool is_try) {
-    assert(! is_try || (t->isA<hilti::type::SignedInteger>() || t->isA<hilti::type::UnsignedInteger>() ||
-                        t->isA<hilti::type::Bitfield>()));
+Expression* ParserBuilder::parseType(UnqualifiedType* t, const production::Meta& meta, Expression* dst, TypesMode mode,
+                                     bool no_trim) {
+    if ( auto e = TypeParser(this, meta, dst, mode).buildParser(t); e || mode == TypesMode::Optimize ) {
+        if ( mode == TypesMode::Default && ! no_trim )
+            trimInput();
```

Comment:
> Don't we need to trim for "optimize" mode as well?

Trimming already happens later for the `optimize` (because it will call `advanceInput()`, so we'd do it twice.

The trimming semantics are a bit of a mess unfortunately because we use `parseType()` from different places with different semantics. At first, I tried removing the trimming from `parseType()` altogether, leaving it up to the caller, but that misaligns with `parseLiteral()`, which performs trimming as well and can't be easily changed. So I went back to letting `parseType()` trim, but with some restrictions. FWIW, our tests pretty nicely catch when something goes wrong.